File: my-folder/0199-binary-tree-right-side-view/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rightSideView(root *TreeNode) []int {
    res := []int{}

    records := [][]int{}
    
    var dfs func(root *TreeNode, depth int)
    dfs = func(root *TreeNode, depth int) {
        if root == nil {
            return
        }
        if len(records) == depth {
            records = append(records, []int{})
        }
        records[depth] = append(records[depth], root.Val)
        dfs(root.Left, depth+1)
        dfs(root.Right, depth+1)
    }


    dfs(root, 0)

    for _, level := range records {
        res = append(res, level[len(level)-1])
    }

    return res
}

File: my-folder/0199-binary-tree-right-side-view/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
    let res: number[] = []
    
    let dfs = function(root: TreeNode | null, depth: number) {
        if (root == null)
            return

        if (res.length === depth) {
            res.push(root.val)
        }
        
        dfs(root.right, depth + 1)
        dfs(root.left, depth + 1)
    }

    dfs(root, 0)
    return res
};
